README.md
[![DOI](https://zenodo.org/badge/5630/hrbrmstr/resolv.png)](http://dx.doi.org/10.5281/zenodo.11343)

resolv
======

ldns DNS resolver wrapper libary for R

Needs `ldns` - <http://www.nlnetlabs.nl/projects/ldns/> - which is `apt`-able and `brew`-able.

    library(devtools)
    install_github("hrbrmstr/resolv")

[These folks](http://dev.telnic.org/trac/wiki/DotTelUtils) seem to have done some work getting the `ldns` library to work under Windows, but this particular package only works (for now on Linux/Mac OS X.

Bug reports (esp from ppl with more C++/Rcpp experience), feature requests & pull requests welcome/encouraged. The code/package is documented pretty well (esp for me). Hopefully this library can replace `system` calls for folks who need to "do DNS stuff" from R.

### News

-   Version update to `0.2.3` removed `plyr` dependency; cleaned up tests; converted alexa CSV to rda file (can now do `data(alexa)`)
-   Version update to `0.2.2` includes making the parameters fully consistent, making the vectorized functions work better and having even saner return values when there were errors or no records found
-   Version update to `0.2.1` includes 2 memory fixes and better return types if no records are found
-   Version update to `0.2.0` includes ability to (optionally - set the `full` parameter to `TRUE`) return `class`, `ttl` & `owner` fields, includes `resolve_ns()` and `NS()` functions, plus changes return type for a few functions.
-   Version update to `0.1.2` after running `valgrind` and fixing some missing `free`'s (`#ty` to [@arj](<http://twitter.com/arj>)!)
-   Version update to `0.1.1` as I modified some of the roxygen documentation to better make this work out of the box. Any help getting it to work on Windows is greatly appreciated

### Description

Provides functions to perform robust DNS lookups from R. Uses the `ldns` library which provides support for IPv4 & IPv6 addresses as well as DNSSEC. This library currently exposes the functions indicated below.

### Details

    Package: resolv
    Type: Package
    Title: Wrapper to ldns library for DNS calls from R
    Version: 0.2.3
    Date: 2014-08-30
    Author: Bob Rudis (@hrbrmstr)
    Maintainer: Bob Rudids <bob@rudis.net>
    Description: Wrapper to ldns library for DNS calls from R. It provides
        vecorized & non-vectorized version of core DNS query functions and
        allows for vector or data frame return values in most functions
        (depending on the level of detail desired from the query). It's
        significantly faster than issuing a system() call but does not
        yet work under Windows.
    URL: https://github.com/hrbrmstr/resolv
    BugReports: https://github.com/hrbrmstr/resolv/issues
    License: GPL-2
    Imports:
        Rcpp (>= 0.11.1)
    LinkingTo: Rcpp

Direct `ldns` wrappers:

-   `resolv_txt()` - perform TXT lookups
-   `resolv_mx()` - perform MX lookups (returns data frame)
-   `resolv_cname()` - perform CNAME lookups
-   `resolv_srv()` - perform SRV lookups (returns data frame)
-   `resolv_a()` - perform A lookups
-   `resolv_ptr()` - perform PTR lookups
-   `resolv_ns()` - perform NS lookups

and, their vectorized counterparts:

-   `TXT()`
-   `MX()`
-   `CNAME()`
-   `SRV()`
-   `A()`
-   `PTR()`
-   `NS()`

(TODO: to add "SOA" suppot)

Ancillary/"fun"ctions

These show off some of what you can do with DNS

-   `ip2asn()` - interface to <http://www.team-cymru.org/Services/ip-to-asn.html#dns>
-   `asninfo()` - interface to <http://www.team-cymru.org/Services/ip-to-asn.html#dns>
-   `wikidns()` - interface to <https://dgl.cx/wikipedia-dns>
-   `dnscalc()` - interface to <http://www.isi.edu/touch/tools/dns-calc.html>

### Installation

``` {.r}
devtools::install_github("hrbrmstr/resolv")
```

> Don't forget the need to `brew install ldns` or your favorite linux pkg mgr equivalent, first.

### Usage

``` {.r}
require(resolv)
```

    ## Loading required package: resolv

``` {.r}
library(plyr)

## google talk provides a good example for this
resolv_srv("_xmpp-server._tcp.gmail.com.")
```

    ##                           fqdn priority weight port                         target
    ## 1 _xmpp-server._tcp.gmail.com.        5      0 5269      xmpp-server.l.google.com.
    ## 2 _xmpp-server._tcp.gmail.com.       20      0 5269 alt1.xmpp-server.l.google.com.
    ## 3 _xmpp-server._tcp.gmail.com.       20      0 5269 alt2.xmpp-server.l.google.com.
    ## 4 _xmpp-server._tcp.gmail.com.       20      0 5269 alt3.xmpp-server.l.google.com.
    ## 5 _xmpp-server._tcp.gmail.com.       20      0 5269 alt4.xmpp-server.l.google.com.

``` {.r}
## where www.nasa.gov hosts
resolv_a("www.nasa.gov")
```

    ## [1] "69.28.157.198"

``` {.r}
## DNS seekrit TXT URLs
## uncomment and run it locally
# browseURL(gsub("\"", "", resolv_txt("google-public-dns-a.google.com")))

# vectorized + full return
TXT(c("stackoverflow.com", "microsoft.com", "apple.com", "google.com"), full=TRUE)
```

    ##                fqdn
    ## 1 stackoverflow.com
    ## 2     microsoft.com
    ## 3     microsoft.com
    ## 4         apple.com
    ## 5        google.com
    ##                                                                                                                                                                                                                                                          txt
    ## 1                                                                                                                                                      "v=spf1 a mx ip4:198.252.206.0/24 ip4:69.59.197.0/26 include:cmail1.com include:_spf.google.com ~all"
    ## 2                                                                                                                                                                 "FbUF6DbkE+Aw1/wi9xgDi8KVrIIZus5v8L6tbIQZkGrQ/rVQKJi8CjQbBtWtE64ey4NJJwj5J65PIggVYNabdQ=="
    ## 3 "v=spf1 include:_spf-a.microsoft.com include:_spf-b.microsoft.com include:_spf-c.microsoft.com include:_spf-ssg-a.microsoft.com include:spf-a.hotmail.com ip4:147.243.128.24 ip4:147.243.128.26 ip4:147.243.128.25 ip4:147.243.1.47 ip4:147.243.1.48 -all"
    ## 4                                                                                                                                                                                                                               "v=spf1 ip4:17.0.0.0/8 ~all"
    ## 5                                                                                                                                                                              "v=spf1 include:_spf.google.com ip4:216.73.93.70/31 ip4:216.73.93.72/31 ~all"
    ##                owner class  ttl
    ## 1 stackoverflow.com.     1  299
    ## 2     microsoft.com.     1  311
    ## 3     microsoft.com.     1  311
    ## 4         apple.com.     1 3177
    ## 5        google.com.     1 3030

``` {.r}
## parallel queries
library(foreach)
library(doParallel)
```

    ## Loading required package: iterators
    ## Loading required package: parallel

``` {.r}
library(data.table)

data(alexa) # http://s3.amazonaws.com/alexa-static/top-1m.csv.zip

n <- 100 # top 'n' to resolve (change it to 10000 on your own as that # makes the README builds churn too long)

registerDoParallel(cores=6) # set to what you can on your system
output <- foreach(i=1:n, .packages=c("Rcpp", "resolv")) %dopar% resolv_a(alexa[i,]$domain)
names(output) <- alexa[1:n,]$domain
head(output)
```

    ## $facebook.com
    ## [1] "173.252.110.27"
    ## 
    ## $youtube.com
    ##  [1] "74.125.226.0"  "74.125.226.1"  "74.125.226.2"  "74.125.226.3"  "74.125.226.4"  "74.125.226.5"  "74.125.226.6" 
    ##  [8] "74.125.226.7"  "74.125.226.8"  "74.125.226.9"  "74.125.226.14"
    ## 
    ## $yahoo.com
    ## [1] "98.138.253.109" "98.139.183.24"  "206.190.36.45" 
    ## 
    ## $baidu.com
    ## [1] "123.125.114.144" "220.181.111.85"  "220.181.111.86" 
    ## 
    ## $wikipedia.org
    ## [1] "208.80.154.224"
    ## 
    ## $twitter.com
    ## [1] "199.16.156.6"   "199.16.156.38"  "199.16.156.198" "199.16.156.230"

### References

-   <http://www.nlnetlabs.nl/projects/ldns/>

### See Also

-   <http://www.nlnetlabs.nl/projects/ldns/>
-   <http://dds.ec/blog/posts/2014/Apr/making-better-dns-txt-record-lookups-with-rcpp/>

### Test Results

``` {.r}
library(resolv)
library(testthat)

date()
```

    ## [1] "Sat Aug 30 15:02:48 2014"

``` {.r}
test_dir("tests/")
```

    ## resolv core : ......


NEWS.md

### News

- Version update to `0.2.3` removed `plyr` dependency; cleaned up tests; converted alexa CSV to rda file (can now do `data(alexa)`)
- Version update to `0.2.2` includes making the parameters fully consistent, making the vectorized functions work better and having even saner return values when there were errors or no records found
- Version update to `0.2.1` includes 2 memory fixes and better return types if no records are found
- Version update to `0.2.0` includes ability to (optionally - set the `full` parameter to `TRUE`) return `class`, `ttl` & `owner` fields, includes `resolve_ns()` and `NS()` functions, plus changes return type for a few functions.
- Version update to `0.1.2` after running `valgrind` and fixing some missing `free`'s (`#ty` to [@arj](http://twitter.com/arj)!)
- Version update to `0.1.1` as I modified some of the roxygen documentation to better make this work out of the box. Any help getting it to work on Windows is greatly appreciated


tests/test-all.R
library(testthat)
test_check("resolv")

tests/testthat/test-resolv.R
context("resolv core")

test_that("we can use the ldns library", {
  
  # success
  expect_that(resolv_a("etsy.com"), is_a("character"))
  expect_that(resolv_txt("google.com"), is_a("character"))
  expect_that(resolv_mx("securitymetrics.org"), is_a("data.frame"))
  expect_that(resolv_cname("www.paypal.com"), is_a("character"))
  expect_that(resolv_ptr("69.28.187.45"), is_a("character"))
  expect_that(resolv_srv("_xmpp-server._tcp.gmail.com."), is_a("data.frame"))
  
})



R/resolv.R
#' Vectorized version of \link{resolv_a}
#' 
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return named vector or list
#' @export
A <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_a, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(f) { resolv_a(f, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_txt}
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return named vector or list
#' @export
TXT <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_txt, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(fqdn) { resolv_txt(fqdn, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_mx}
#'
#' @param fqdn input character vector (domain name)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return named vector or list 
#' @export
MX <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_mx, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(fqdn) { resolv_mx(fqdn, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_cname}
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return list
#' @export
CNAME <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_cname, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(fqdn) { resolv_cname(fqdn, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_ns}
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return list
#' @export
NS <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_ns, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(fqdn) { resolv_ns(fqdn, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_ptr}
#'
#' @param ip address input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return list
#' @export
PTR <- function(ip, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(ip, resolv_ptr, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(ip, function(ip) { resolv_ptr(ip, nameserver, showWarnings, full) }))
  }
  
}

#' Vectorized version of \link{resolv_srv}
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return list
#' @export
SRV <- function(fqdn, nameserver=NA_character_, showWarnings=FALSE, full=FALSE) {
  
  if (full) {
    return(Reduce(rbind, lapply(fqdn, resolv_srv, nameserver=nameserver, showWarnings=showWarnings, full=full)))
  } else {
    return(sapply(fqdn, function(fqdn) { resolv_srv(fqdn, nameserver, showWarnings, full) }))
  }
  
}

#' Return ASN info from Team CYNRU DNS lookup service
#' 
#' Pretty much provided as an example of data services. You
#' should use the one in the \link[iptools]{iptools} or \link[netintel]{netintel} packages
#' if you are serious about doing IP/ASN lookups.
#' 
#' @param ip address to lookup (character vector)
#' @return data frame containing named ASN attributes
#' @export
ip2asn <- function(ip="216.90.108.31") {
  
  Reduce(rbind, lapply(ip, function(ip) {
    
    orig <- ip
    
    ip <- paste(paste(rev(unlist(strsplit(ip, "\\."))), sep="", collapse="."), 
                ".origin.asn.cymru.com", sep="", collapse="")
    result <- resolv_txt(ip)
    if (length(result) == 0) {
      return(data.frame(ip=orig, asn=NA, cidr=NA, cn=NA, registry=NA, regdate=NA))
    }
    out <- unlist(strsplit(gsub("\"", "", result), "\ *\\|\ *"))
    
    return(data.frame(ip=orig, asn=out[1], cidr=out[2], cn=out[3], registry=out[4], regdate=out[5]))
    
  }))
  
}

#' Return ASN info from Team CYNRU DNS lookup service
#' 
#' @param asn number (with or without "AS" prefixed) to lookup (character vector)
#' @return data frame containing named ASN attributes
#' @export
asninfo <- function(asn="AS23028") {

  Reduce(rbind, lapply(asn, function(asn) {
    
    orig <- asn
    
    # prefix with "AS" in case it isn't
    asn <- gsub("^([0-9]+)", "AS\\1", asn)
    asn <- paste(asn, ".asn.cymru.com", sep="", collapse="")
    result <- resolv_txt(asn)
    if (length(result) == 0) {
      return(data.frame(asn=orig, cn=NA, registry=NA, regdate=NA, location=NA))
    }
    out <- unlist(strsplit(gsub("\"", "", result), "\ *\\|\ *"))
    
    return(data.frame(asn=out[1], cn=out[2], registry=out[3], regdate=out[4], location=out[5]))
    
  }))
  
}

#' Return wikipedia topic via DNS
#' 
#' @param term wikipedia term to lookup
#' @return vector of TXT record results for term description
#' @export
#' @seealso https://dgl.cx/wikipedia-dns
wikidns <- function(term="bayes") {
  term <- paste(term, ".wp.dg.cx", sep="", collapse="")
  return(resolv_txt(term))
  
}

#' Return simple math computations via DNS
#' 
#' @param a first number (0-99)
#' @param b second number (0-99)
#' @param op calculation to perform (add|sub|mul|div)
#' @return result (in IP address octet notation)
#' @export
#' @seealso http://www.isi.edu/touch/tools/dns-calc.html
dnscalc <- function(a=0:99, b=0:99, op=c("add", "sub", "mul", "div")) {
  op <- ifelse(op %in% c("add", "sub", "mul", "div"), op, "add")
  calc <- paste(a%%100, ".", b%%100, ".", op, ".calc.postel.org", sep="", collapse="")
  return(resolv_a(calc, "dns.postel.org"))
}

#' Wrapper to ldns library for DNS calls from R
#' @docType package
#' @name resolv-package
#' @aliases resolv
#' @author Bob Rudis <bob@@rudis.net>
#' @import Rcpp
#' @useDynLib resolv
#' @references \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso
#' \itemize{
#'   \item \url{https://github.com/hrbrmstr/resolv}
#'   \item \url{http://www.nlnetlabs.nl/projects/ldns/}
#'   \item \url{http://dds.ec/blog/posts/2014/Apr/making-better-dns-txt-record-lookups-with-rcpp/}
#' }
#' @examples
#' \dontrun{
#' require(resolv)
#'
#' ## google talk provides a good example for this
#' resolv_srv("_xmpp-server._tcp.gmail.com.")
#' 
#' ##                           fqdn priority weight port                         target
#' ## 1 _xmpp-server._tcp.gmail.com.        5      0 5269      xmpp-server.l.google.com.
#' ## 2 _xmpp-server._tcp.gmail.com.       20      0 5269 alt1.xmpp-server.l.google.com.
#' ## 3 _xmpp-server._tcp.gmail.com.       20      0 5269 alt2.xmpp-server.l.google.com.
#' ## 4 _xmpp-server._tcp.gmail.com.       20      0 5269 alt3.xmpp-server.l.google.com.
#' ## 5 _xmpp-server._tcp.gmail.com.       20      0 5269 alt4.xmpp-server.l.google.com. 
#' 
#' ## where www.nasa.gov hosts
#' resolv_a("www.nasa.gov", full=TRUE)
#' 
#' ##           fqdn       address                owner class ttl
#' ## 1 www.nasa.gov 69.28.157.198 iznasa.hs.llnwd.net.     1 274
#'
#' resolv_ptr("69.28.187.45")
#' ## [1] "cds355.iad.llnw.net."
#' 
#' ## vectorized
#' A(c("dds.ec", "rud.is"), full=TRUE)
#' 
#' ## $dds.ec
#' ##     fqdn       address   owner class ttl
#' ## 1 dds.ec 162.243.111.4 dds.ec.     1  87
#' ## 
#' ## $rud.is
#' ##     fqdn        address   owner class   ttl
#' ## 1 rud.is 184.106.97.102 rud.is.     1 17781
#
#' # seekrit URLs
#' browseURL(gsub("\"", "", resolv_txt("google-public-dns-a.google.com")))
#'
#' }
NULL

#' @title Alexa Top 1 Million Sites
#' @description Alexa Top 1 Million Sites
#' \itemize{
#'   \item \code{rank}. rank of the domain (int)
#'   \item \code{domain}. the domain name (chr)
#' }
#'
#' @docType data
#' @keywords datasets
#' @name alexa
#' @seealso \itemize{
#'   \item IANA - \url{http://aws.amazon.com/alexa-top-sites/}
#' }
#' @usage data(alexa)
#' @note Last updated 2014-08-09
#' @format A data frame with 1,000,024 rows and 2 variables
NULL


R/RcppExports.R
# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Returns the DNS A records for a given FQDN
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return vector or data frame (if \code{full}==\code{TRUE}) of A records or \code{character(0)} if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/} (cool DNS A hacks vla \url{https://twitter.com/habbie/status/460067198586081280})
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#' 
#' ## single address return
#' resolv_a("etsy.com")
#' [1] "38.106.64.123"
#' 
#' ## multiple address returns
#' resolv_a("google.com")
#' [1] "173.194.43.0"  "173.194.43.1"  "173.194.43.2"  "173.194.43.3" 
#' [5] "173.194.43.4"  "173.194.43.5"  "173.194.43.6"  "173.194.43.7" 
#' [9] "173.194.43.8"  "173.194.43.9"  "173.194.43.14"
#' 
#' ## must put at least one DNS hack in
#' resolv_a("10.15.add.calc.postel.org", "dns.postel.org")
#' [1] "0.25.0.0"
#' }
resolv_a <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_a', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

#' Returns the DNS TXT records for a given FQDN
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return vector or data frame (if \code{full}==\code{TRUE}) of TXT records or \code{character(0)} if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/} (cool DNS TXT hacks vla \url{https://twitter.com/habbie/status/460067198586081280})
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#' 
#' ## get the TXT record for Google
#' resolv_txt("google.com")
#' [1] "\"v=spf1 include:_spf.google.com ip4:216.73.93.70/31 ip4:216.73.93.72/31 ~all\""
#' 
#' ## Use Team CYMRU ASN TXT record lookup service
#' resolv_txt("31.108.90.216.origin.asn.cymru.com")
#' [1] "\"23028 | 216.90.108.0/24 | US | arin |\""
#'
#' ## Wikipedia lookups:
#' resolv_txt("bind.wp.dg.cx")
#' [1] "\"BIND , or named , is the most widely used DNS software on the Internet. On Unix-like operating systems it is the de facto standard. http://en.wikipedia.org/wiki/BIND\""
#'
#' ## seekrit URLs
#' browseURL(gsub("\"", "", resolv_txt("google-public-dns-a.google.com")))
#' 
#' ## get the TXT records for PayPal
#' resolv_txt("paypal.com")
#' [1] "\"MS=ms96239109\""                                                                                                                                                                                        
#' [2] "\"yandex-verification: 73acb90f6a9abd76\""                                                                                                                                                                
#' [3] "\"google-site-verification=NrhK1Hj7KuCPua1OcvfacDawt46H9VjByS4IAw5vsFA\""                                                                                                                                 
#' [4] "\"v=spf1 include:pp._spf.paypal.com include:3rdparty._spf.paypal.com include:3rdparty1._spf.paypal.com include:3rdparty2._spf.paypal.com include:3rdparty3._spf.paypal.com include:c._spf.ebay.com ~all\""
#' }
resolv_txt <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_txt', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

#' Returns the DNS MX records for a given domain
#'
#' @param fqdn input character vector (domain name)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return data frame of MX records (preference & exchange; +owner,class,ttl if \code{full}==\code{TRUE}) or an empty data frame if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/} (cool DNS MX hacks vla \url{https://twitter.com/habbie/status/460067198586081280})
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#' 
#' resolv_mx("rudis.net", full=TRUE)
#' ##        fqdn prefernece                 exchange      owner class ttl
#' ## 1 rudis.net          1      aspmx.l.google.com. rudis.net.     1 599
#' ## 2 rudis.net          5 alt1.aspmx.l.google.com. rudis.net.     1 599
#' ## 3 rudis.net          5 alt2.aspmx.l.google.com. rudis.net.     1 599
#' ## 4 rudis.net         10   aspmx2.googlemail.com. rudis.net.     1 599
#' ## 5 rudis.net         10   aspmx3.googlemail.com. rudis.net.     1 599
#' ## 6 rudis.net         10   aspmx4.googlemail.com. rudis.net.     1 599
#' ## 7 rudis.net         10   aspmx5.googlemail.com. rudis.net.     1 599
#' ## 8 rudis.net        100  mx-caprica.easydns.com. rudis.net.     1 599
#' }
resolv_mx <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_mx', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

#' Returns the DNS CNAME records for a given FQDN
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return vector or data frame (if \code{full}==\code{TRUE}) of CNAME records or \code{character(0)} if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/}
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#'
#' resolv_cname("www.paypal.com")
#' [1] "www.paypal.com.akadns.net."
#' }
resolv_cname <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_cname', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

#' Returns the DNS NS records for a given FQDN
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return vector or data frame (if \code{full}==\code{TRUE}) of NS records or \code{character(0)} if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/}
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#'
#' resolv_ns("www.paypal.com")
#' }
resolv_ns <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_ns', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

#' Returns the DNS PTR records for a given IP address
#'
#' @param ip address input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return vector or data frame (if \code{full}==\code{TRUE}) of PTR records or \code{character(0)} if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/}
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#'
#' # where www.nasa.gov hosts
#' resolv_a("www.nasa.gov")
#' [1] "69.28.187.45"    "208.111.161.110"
#' 
#' resolv_ptr("69.28.187.45")
#' [1] "cds355.iad.llnw.net."
#' 
#' ## big one - truncated output
#' resolv_ptr("17.149.160.49")
#'   [1] "asto.re."                   "next.com."                 
#'   [3] "qtml.com."                  "qttv.net."                 
#'   [5] "apple.com."                 "apple.info."               
#'   [7] "ikids.com."                 "qt-tv.net."                
#'   [9] "carbon.com."                "eworld.com."
#' ...  
#' [131] "quicktimestreaming.net."    "publishing-research.com."  
#' [133] "publishing-research.org."   "applefinalcutproworld.com."
#' [135] "applefinalcutproworld.net." "applefinalcutproworld.org."
#' }
resolv_ptr <- function(ip, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_ptr', PACKAGE = 'resolv', ip, nameserver, showWarnings, full)
}

#' Returns the DNS SRV records for a given FQDN
#'
#' @param fqdn input character vector (FQDN)
#' @param nameserver the nameserver to send the request to (optional; uses standard resolver behavior if not specified)
#' @param showWarnings display R warning messages (bool)
#' @param full include full record response information in results (bool)
#' @return data frame of SRV records (named fields; +owner,class,ttl if \code{full}==\code{TRUE}) or an empty list if none
#' @seealso \url{http://www.nlnetlabs.nl/projects/ldns/}
#' @seealso \url{http://www.cambus.net/interesting-dns-hacks/}
#' @export
#' @examples
#' \dontrun{
#' require(resolv)
#' library(plyr)
#' 
#' ## google talk provides a good example for this
#' resolv_srv("_xmpp-server._tcp.gmail.com.", full=TRUE)
#' ##                           fqdn priority weight port                         target                        owner class ttl
#' ## 1 _xmpp-server._tcp.gmail.com.        5      0 5269      xmpp-server.l.google.com. _xmpp-server._tcp.gmail.com.     1 804
#' ## 2 _xmpp-server._tcp.gmail.com.       20      0 5269 alt1.xmpp-server.l.google.com. _xmpp-server._tcp.gmail.com.     1 804
#' ## 3 _xmpp-server._tcp.gmail.com.       20      0 5269 alt2.xmpp-server.l.google.com. _xmpp-server._tcp.gmail.com.     1 804
#' ## 4 _xmpp-server._tcp.gmail.com.       20      0 5269 alt3.xmpp-server.l.google.com. _xmpp-server._tcp.gmail.com.     1 804
#' ## 5 _xmpp-server._tcp.gmail.com.       20      0 5269 alt4.xmpp-server.l.google.com. _xmpp-server._tcp.gmail.com.     1 804
#' }
resolv_srv <- function(fqdn, nameserver = NA_character_, showWarnings = FALSE, full = FALSE) {
    .Call('resolv_resolv_srv', PACKAGE = 'resolv', fqdn, nameserver, showWarnings, full)
}

